rhel/repositoryscanner.go

```diff
@@ -0,0 +1,263 @@
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime/trace"
+	"strings"
+	"time"
+
+	"github.com/docker-slim/docker-slim/pkg/docker/dockerfile/ast"
+	"github.com/rs/zerolog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/internal/indexer"
+	"github.com/quay/claircore/pkg/cpe"
+)
+
+// RepositoryScanner implements Red Hat Cpes based repositories
+type RepositoryScanner struct {
+	cpeFetcher *containerAPI
+	timeout    time.Duration
+}
+
+// RepoScannerConfig is the struct that will be passed to
+// (*RepositoryScanner).Configure's ConfigDeserializer argument.
+type RepoScannerConfig struct {
+	Timeout time.Duration `json:"timeout",yaml:"timeout"`
+	API     string        `json:"api",yaml:"api"`
+}
+
+// RedHatCPERepositoryKey is a key of Red Hat's CPE based repository
+const RedHatCPERepositoryKey = "rhel-cpe-repo"
+
+// Name implements scanner.Name.
+func (*RepositoryScanner) Name() string { return "rhel-cpe-scanner" }
+
+// Version implements scanner.VersionedScanner.
+func (*RepositoryScanner) Version() string { return "1.0" }
+
+// Kind implements scanner.VersionedScanner.
+func (*RepositoryScanner) Kind() string { return "repository" }
+
+const DefaultContainerAPI = "https://catalog.redhat.com/api/containers/"
+
+// Configure implements the RPCScanner interface.
+func (r *RepositoryScanner) Configure(ctx context.Context, f indexer.ConfigDeserializer, c *http.Client) error {
+	cfg := RepoScannerConfig{}
+	if err := f(&cfg); err != nil {
+		return err
+	}
+	// Set defaults if not set via passed function.
+	if cfg.Timeout == 0 {
+		cfg.Timeout = 30 * time.Second
+	}
+	if cfg.API == "" {
+		cfg.API = DefaultContainerAPI
+	}
+
+	root, err := url.Parse(cfg.API)
+	if err != nil {
+		return err
+	}
+
+	r.cpeFetcher = &containerAPI{
+		root:   root,
+		client: c,
+	}
+	r.timeout = cfg.Timeout
+	return nil
+}
+
+// Scan gets Red Hat repositories based on CPE information.
+func (r *RepositoryScanner) Scan(ctx context.Context, l *claircore.Layer) (repositories []*claircore.Repository, err error) {
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "rhel/RepositoryScanner.Scan").
+		Str("version", r.Version()).
+		Str("layer", l.Hash.String()).
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Debug().Msg("start")
+	defer log.Debug().Msg("done")
+
+	cpes, err := contentSets(l)
+	if err != nil {
+		return []*claircore.Repository{}, err
+	}
+	if len(cpes) == 0 && r.cpeFetcher != nil {
+		// Embedded content-sets are available only for new images.
+		// For old images, use fallback option and query Red Hat Container API.
+		ctx, done := context.WithTimeout(ctx, r.timeout)
+		defer done()
+		cpes, err = r.containerAPI(ctx, &log, l)
+		if err != nil {
+			return []*claircore.Repository{}, err
+		}
+	}
+
+	for _, n := range cpes {
+		r := &claircore.Repository{
+			Name: n,
+			Key:  RedHatCPERepositoryKey,
+		}
+		r.CPE, err = cpe.Unbind(n)
```

Comment:
Repository CPE property is also missing in the index report.

`"repository": {
    "1": {
      "id": "1",
      "name": "cpe:/a:redhat:enterprise_linux:8::appstream",
      "key": "rhel-cpe-repo",
      "cpe": ""
    },
    "2": {
      "id": "2",
      "name": "cpe:/o:redhat:enterprise_linux:8::baseos",
      "key": "rhel-cpe-repo",
      "cpe": ""
    }
  },`